components/shared/code/libs/LIB_nvm.c

```diff
@@ -0,0 +1,517 @@
+/** LIB_nvm.c
+ * 	Source code for the NVM Manager
+ */
+
+/******************************************************************************
+ *                             I N C L U D E S
+ ******************************************************************************/
+
+#include "LIB_nvm.h"
+#include <stdint.h>
+#if FEATURE_IS_ENABLED(NVM_LIB_ENABLED)
+#include "LIB_utility.h"
+#include "libcrc.h"
+#include "string.h"
+
+/******************************************************************************
+ *                              D E F I N E S
+ ******************************************************************************/
+
+#define LIB_NVM_VERSION 1U
+#define NVM_TOTAL_ENTRYID_COUNT (NVM_ENTRYID_INTERNAL_COUNT + COUNTOF(lib_nvm_entries))
+
+#if FEATURE_IS_ENABLED(NVM_FLASH_BACKED) && (FEATURE_IS_DISABLED(MCU_STM32_PN) == false)
+#define SET_STATE 0x00
+#else
+#error "Only flash backed nvm supported presently."
+#endif
+
+/******************************************************************************
+ *                              E X T E R N S
+ ******************************************************************************/
+
+extern const lib_nvm_entry_S lib_nvm_entries[NVM_ENTRYID_COUNT];
+
+#if FEATURE_IS_ENABLED(NVM_FLASH_BACKED)
+extern storage_t __FLASH_NVM_ORIGIN;
+extern storage_t __FLASH_NVM_END;
+storage_t * const nvm_origin = &__FLASH_NVM_ORIGIN;
+storage_t * const nvm_end = &__FLASH_NVM_END;
+#endif
+
+/******************************************************************************
+ *                             T Y P E D E F S
+ ******************************************************************************/
+
+typedef uint8_t lib_nvm_crc_t;
+
+typedef enum
+{
+	NVM_ENTRYID_INTERNAL_LOG = 0U,
+	NVM_ENTRYID_INTERNAL_CYCLE, // Anytime an entry is added, a spare should be removed
+	NVM_ENTRYID_INTERNAL_SPARE_1,
+	NVM_ENTRYID_INTERNAL_SPARE_2,
+	NVM_ENTRYID_INTERNAL_SPARE_3,
+	NVM_ENTRYID_INTERNAL_COUNT,
+} lib_nvm_entryIdInternal_E;
+#define NVM_LAST_USED_INTERNAL_ENTRYID NVM_ENTRYID_INTERNAL_CYCLE
+
+typedef struct
+{
+	lib_nvm_entry_t entryId;
+    storage_t entry_version;
+	lib_nvm_crc_t crc;
+	storage_t initialized;
+	storage_t discarded;
```

Comment:
Store size of record in the header so we dont lose an entire block if a record is currpt/invalid/dogshit